### PR TITLE
Create consolidated booknotes page

### LINKED ISSUE #4868: Create page where all of a patron's book notes are listed
### Describe the problem that you'd like solved

All logged in patrons should be able to view all of the notes and observations that they have submitted on a single page.

### Proposal & Constraints

The new consolidated book notes page will have a similar layout to the reading log page.  The page will have a paginated view containing a list of book notes entries.  Each entry will contain the book's title, a cover image (if available), and any notes and observations that have been left for the book.

Patrons will be given the option to delete an entries notes or observations.  They will also be able to update the same.  Updates to an entry will occur within the existing book notes modal.

### Additional context

This feature will build upon the work done for #4867.

### PR DESCRIPTION
**Pull request by jamesachamp**
Closes #4868

Creates a page where a patron can view and update all of their book notes and observations.  

### Technical

Book notes page can be found at `/account/books/notes`.  A new link in the account drop-down links to this page.  The page contains a paginated view of notes and observations list items.  Up to 25 works can be displayed per page (this may be a bit too many, as notes for multiple editions can be present for a single work).  Each work list item contains tabs that are used to switch between notes and observations.

Notes and observations can be updated or deleted.  Notes can only be updated from the notes tab view, while observations are updated via the book notes modal.  The modals for this page have been modified and do not contain the section that allows a patron to submit notes (this section is still present in the modal for book pages).  Were this section still present, it would be difficult to update the view after a note was submitted.

If observations are deleted after the book notes modal is opened, and the modal is opened again, the form is repopulated with information from the database.  If it wasn't, the deleted values would still appear to be selected when the modal was opened.  As I'm writing this, I'm thinking that it may be better to simply uncheck all values via Javascript on observation delete...

A toast component was created to provide feedback when a note is updated.  This component is stickied near the top of the screen, and is removed after 5 seconds (it can also be closed manually by patrons).  When observations are updated and the modal is closed, the [throbber](https://github.com/internetarchive/openlibrary/wiki/Design-Pattern-Library#throbber) is displayed in the observation view and updated observations are fetched from the database. 

### Screenshot

booknotes_preview

### Stakeholders

@mekarpeles 
@cdrini

### CODE REVIEW

File: openlibrary/plugins/upstream/account.py [hunk 1]
```diff
@@ -724,6 +726,83 @@ def get_works(self, key, page=1, limit=RESULTS_PER_PAGE):
...
+
+        return notes_and_observations
+
+    def get_work(self, work_key):
+        work = web.ctx.site.get(work_key)
+        author_keys = [a.author.key for a in work.get('authors', [])]
+
+        return {
+            'cover_url': (
+                work.get_cover_url('S') or
```
Comment (jamesachamp) [inline openlibrary/plugins/upstream/account.py:774]: Is this a correct way to get a work's cover image?  I'm not able to import works locally and images for the default works seem to be fetched using IA IDs.

File: openlibrary/plugins/openlibrary/js/Toast.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
+import $ from 'jquery';
+
+const DEFAULT_TIMEOUT = 5000;
+
+
+export class Toast {
```
Comment (mekarpeles) [inline openlibrary/plugins/openlibrary/js/Toast.js:6]: Is Toast.js a testing class? Is this related to our modals?

File: openlibrary/plugins/upstream/account.py (hunk 1 above)
Comment (mekarpeles) [reply openlibrary/plugins/upstream/account.py:774]: I think this should get us what we want -- I'll throw on testing so we can see the behavior

File: openlibrary/plugins/openlibrary/js/Toast.js (hunk 2 above)
Comment (mekarpeles) [reply openlibrary/plugins/openlibrary/js/Toast.js:6]: > A toast component was created to provide feedback when a note is updated. This component is stickied near the top of the screen, and is removed after 5 seconds (it can also be closed manually by patrons). When observations are updated and the modal is closed, the throbber is displayed in the observation view and updated observations are fetched from the database.

never mind, I see! TIL: toast messages :)

Comment (jamesachamp) [reply openlibrary/plugins/openlibrary/js/Toast.js:6]: That's what they are called in Android.  I'm not sure if this type of component is known as a toast message in a web development context, though...

I've noticed that these toast components are displacing other elements on the page, so I'll need to figure out how to keep that from happening.  I also think that a target element in which to mount the toast message to should be passed in the constructor.  This way, we can be sure that the toast message is visible when the modal is open (by mounting it to the modal).

If I can't figure out how to display the messages without disrupting the page layout, I may remove this entirely and provide some other type of feedback for note submissions (maybe replacing the submit button's text with a spinner while waiting for a response from the server).

**Comment by jamesachamp** [comment]
Updated Toast component, allowing it to be displayed while the modal is open:

toast_update

File: openlibrary/templates/account/notes.html [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    <div style="margin-top: 1em;">
+      $if notes_and_observations:
+        <ul>
+            $for i in notes_and_observations:
+              $ work_id = i.work_id
+              $ work_cover_url = i.work['cover_url']
+              
+              <li style="border-bottom: 1px solid #f2f0e7; margin-bottom: 1em; background-color: #f9f9f9;">
+                <div style="display:flex; padding: 1em;">
+                  <span class="imageLg" style="padding-right: 1em;">
```
Comment (jamesachamp) [inline openlibrary/templates/account/notes.html:23]: Inline styling should be removed.

**Comment by jamesachamp** [comment]
Added link to book notes page in reading log:

booknotes-reading-log

**Comment by jamesachamp** [comment]
Removed dedicated book notes page.  All book notes are now viewable at `/account/books/book-notes`.

updated-book-notes-view

File: openlibrary/core/booknotes.py [hunk 4]
```diff
@@ -146,3 +146,110 @@ def remove(cls, username, work_id, edition_id=NULL_EDITION_VALUE):
...
+        only counted once.
+
+        return: A count of distinct work IDs for the given username
+        """
+        oldb = db.get_db()
+        data = {
+            'username': username
+        }
+        query = """
+            SELECT COUNT(DISTINCT(t.work_id)) FROM (
```
Comment (jamesachamp) [inline openlibrary/core/booknotes.py:248]: I'm not sure if `DISTINCT` is needed here.  The `UNION` should only be selecting distinct values.

File: openlibrary/macros/UserMetadata.html [hunk 5]
```diff
@@ -1,25 +1,37 @@
-$def with (work, edition, link_text, id, classes='')
+$def with (link_text, id, classes='', work=None, edition=None, notes=None, work_key=None, reload_id=None)
```
Comment (jamesachamp) [inline openlibrary/macros/UserMetadata.html:1]: Maybe work and edition can be combined as `page` (or `doc`).  Then we can determine if the object is an edition or a work.

File: openlibrary/plugins/openlibrary/js/patron-metadata/index.js [hunk 6]
```diff
@@ -118,39 +123,200 @@ export function initPatronMetadata() {
...
+
+                $.ajax({
+                    type: 'GET',
+                    url: '/observations',
+                    dataType: 'json'
+                })
+                    .done(function(data) {
+                        populateForm($(`#${context.id}-user-metadata`), data.observations, selectedValues, context.id, i18nStrings);
+                        addChangeListeners(context);
+                        $(`#${context.id}-cancel-submission`).click(function() {
```
Comment (jamesachamp) [inline openlibrary/plugins/openlibrary/js/patron-metadata/index.js:156]:
```suggestion
                        $(`#${context.id}-cancel-submission`).on('click', function() {
```

File: openlibrary/plugins/upstream/account.py [hunk 7]
```diff
@@ -739,6 +741,62 @@ def get_works(self, key, page=1, limit=RESULTS_PER_PAGE):
             raise
 
 
+class PatronBookNotes(object):
+    """ Manages the patron's book notes and observations """
+
+    def __init__(self, user=None, limit=RESULTS_PER_PAGE, page=1):
```
Comment (jamesachamp) [inline openlibrary/plugins/upstream/account.py:747]: `RESULTS_PER_PAGE` is 25, which may be too many booknotes to display at once.

**Comment by mekarpeles** [comment]
Tons of components here are great. @jamesachamp and I talked about breaking this page up into two pages to improve performance and user experience:
- My Book Notes
- My Observations

we're hoping this simplified things. Going to close this PR, looking forward to seeing much of the same code in the followup! ty!
